win32ss/user/ntuser/focus.c

```diff
@@ -614,17 +614,24 @@ co_IntSendActivateMessages(PWND WindowPrev, PWND Window, BOOL MouseActivate, BOO
 VOID FASTCALL
 IntSendFocusMessages( PTHREADINFO pti, PWND pWnd)
 {
-   PWND pWndPrev;
+   PWND pWndPrev, pRealWndPrev;
    PUSER_MESSAGE_QUEUE ThreadQueue = pti->MessageQueue; // Queue can change...
    HWND hwndPrev;
+   USER_REFERENCE_ENTRY Ref;
 
    ThreadQueue->QF_flags &= ~QF_FOCUSNULLSINCEACTIVE;
    if (!pWnd && ThreadQueue->spwndActive)
    {
       ThreadQueue->QF_flags |= QF_FOCUSNULLSINCEACTIVE;
    }
 
-   pWndPrev = ThreadQueue->spwndFocus;
+   pRealWndPrev = pWndPrev = ThreadQueue->spwndFocus;
+   if (pRealWndPrev)
+   {
+      UserRefObjectCo(pRealWndPrev, &Ref);
+      if (!ValidateHwndNoErr(UserHMGetHandle(pRealWndPrev)))
```

Comment:
Why is that handle stuff needed? Shouldn't it be that from within running win32k code, all these stored non-NULL PWNDs are automatically valid?

Reply:
`spwndFocus` might be invalid.

Reply:
In what circumstances? I would expect it to be either NULL or a pointer to a valid PWND ?

Reply:
Ideally `spwndFocus` should be assignment-locked but ReactOS has different impl.

Reply:
This function is called from `co_UserSetFocus`.

Reply:
The below code sending {{WM_KILLFOCUS}}.

Reply:
This code was useless. Deleted.

Reply:
This code was useless. Deleted.

Reply:
`co_` functions do a callback to usermode code, so the entire world can change during those ;)